sunkit_image/granule.py

```diff
@@ -0,0 +1,298 @@
+"""
+import logging
+
+import numpy as np
+import scipy.ndimage as sndi
+import skimage
+from sklearn.cluster import KMeans
+
+import sunpy
+import sunpy.map
+
+__all__ = [
+    "segment",
+    "get_threshold",
+    "trim_intergranules",
+    "mark_brightpoint",
+    "kmeans_segment",
+    "correlation",
+]
+
+METHODS = ["li", "otsu", "isodata", "mean", "minimum", "yen", "triangle"]
+
+
+def segment(smap, resolution, *, skimage_method="li", mark_dim_centers=False, bp_min_flux=None):
+    """
+    Segment an optical image of the solar photosphere into tri-value maps with:
+
+     * 0 as intergranule
+     * 1 as granule
+     * 1.5 as brightpoint 
+
+    If mark_dim_centers is set to True, an additional label, 0.5, will be assigned to
+    dim grnanule centers. 
+
+    Parameters
+    ----------
+    smap : `~sunpy.map.GenericMap`
+        `~sunpy.map.GenericMap` containing data to segment.
+    resolution : `float`
+        Spatial resolution (arcsec/pixel) of the data.
+    skimage_method : {"li", "otsu", "isodata", "mean", "minimum", "yen", "triangle"}, optional
+        scikit-image thresholding method, defaults to "li".
+        Depending on input data, one or more of these methods may be
+        significantly better or worse than the others. Typically, 'li', 'otsu',
+        'mean', and 'isodata' are good choices, 'yen' and 'triangle' over-
+        identify intergranule material, and 'minimum' over identifies granules.
+    mark_dim_centers : `bool`, optional
+        Whether to mark dim granule centers as a separate category for future exploration.
+    bp_min_flux : `float`, optional
+        Minimum flux per pixel for a region to be considered a brightpoint.
+        Default is `None` which will use data mean + 0.5 * sigma.
+
+    Returns
+    -------
+    segmented_map : `~sunpy.map.GenericMap`
+        `~sunpy.map.GenericMap` containing a segmented image (with the original header).
+    """
+    if not isinstance(smap, sunpy.map.mapbase.GenericMap):
+        raise TypeError("Input must be an instance of a sunpy.map.GenericMap")
+    if skimage_method not in METHODS:
+        raise TypeError("Method must be one of: " + ", ".join(METHODS))
+
+    median_filtered = sndi.median_filter(smap.data, size=3)
+    # Apply initial skimage threshold.
+    threshold = get_threshold(median_filtered, skimage_method)
+    segmented_image = np.uint8(median_filtered > threshold)
+    # Fix the extra intergranule material bits in the middle of granules.
+    seg_im_fixed = trim_intergranules(segmented_image, mark=mark_dim_centers)
+    # Mark brightpoint and get final granule and brightpoint count.
+    seg_im_markbp, brightpoint_count, granule_count = mark_brightpoint(seg_im_fixed, smap.data, resolution, bp_min_flux)
+    logging.info(f"Segmentation has identified {granule_count} granules and {brightpoint_count} brightpoint")
+    segmented_map = sunpy.map.Map(seg_im_markbp, smap.meta)

```

Comment:
Thank you so much for the comments and feedback!

A few questions:

-  What is the data manager? How would I use that?
- I agree it likley makes more sense to just read the resolution from the map. However, I'm not sure I understand what you mean about metadata not being correct for certain maps. What is AIA 304?

Reply:
> * What is the data manager? How would I use that?

This is a utility within sunpy that we can use to manage the test data:
https://docs.sunpy.org/en/stable/generated/api/sunpy.data.data_manager.DataManager.html
and a guide here: https://docs.sunpy.org/en/latest/dev_guide/contents/remote_data.html

We can worry about this after the other items are taken care of. 

> * I agree it likley makes more sense to just read the resolution from the map. However, I'm not sure I understand what you mean about metadata not being correct for certain maps.

I am not sure I will get this correct (Will will correct me) but some metadata is tied directly to the data, so if you replace the data underneath, the metadata is now incorrect and misleading.

In Will's example, 304 is a wavelength and that metadata is in the map.
So if you replace the data with a segmented map, the wavelength keyword is now lying since this data now has no wavelength.

> What is AIA 304?

It is an instrument onboard a satellite that observes the Sun at 304 Angstroms.

Reply:
AIA 304 refers to a full-disk image from the Atmospheric Imaging Assembly (on the Solar Dynamics Observatory - aka SDO) taken with a filter centered on a wavelength of 304 Å.
I think Will's point was that if you start with a map of intensities at a wavelength of 304 Å, but then convert that map to segments,  the segmented map metadata will still say that the segmented map represents intensities at 304 Å, when that is no longer really the case. 
It's tricky, because some of the original map metadata might be useful to propagate (time, coordinates, origin) but not all.

Reply:
Hmm.. I see what you are saying. I guess the issue is that currently the map we return from `segment` takes the header from the input map. I didn't realize that some maps could contain metadata specifying what the pixel values are supposed to represent, as opposed to just specifying the scale, coordinates, etc. I suppose the solution is to read into the new header only the relevant information. I am still quite unfamiliar with SunPy maps, so this would take me a while to figure out.

Reply:
I think an answer to this question would need to come from us. 

The quickest solution would be to create a new header using https://docs.sunpy.org/en/stable/generated/api/sunpy.map.header_helper.make_fitswcs_header.html#sunpy.map.header_helper.make_fitswcs_header

The others would need to chime in here.

Reply:
Sorry for the confusion my comments created! Yes, Kevin's reply perfectly sums up what I meant. 

The simplest way to get around this would be just to create the segmented map using the WCS of the input map as that contains all the things you would want to preserve from the original map (pointing, scale, date, rotation, etc.). This would just look like `Map(seg_data, snap.wcs)`.

Alternatively, if you needed to propagate additional keywords, you could as Nabil suggested use the header creation functionality provided in SunPy.